pyteal/ast/router.py

```diff
@@ -0,0 +1,547 @@
+class OCActions:
+    """
+    OnCompletion Actions keep track of bare-call registrations to all OnCompletion cases.
+    """
+
+    close_out: OCAction = field(kw_only=True, default=OCAction.never())
+    clear_state: OCAction = field(kw_only=True, default=OCAction.never())
+    delete_application: OCAction = field(kw_only=True, default=OCAction.never())
+    no_op: OCAction = field(kw_only=True, default=OCAction.never())
+    opt_in: OCAction = field(kw_only=True, default=OCAction.never())
+    update_application: OCAction = field(kw_only=True, default=OCAction.never())
+
+    def dictify(self) -> dict[EnumInt, OCAction]:
+        return {
+            OnComplete.CloseOut: self.close_out,
+            OnComplete.ClearState: self.clear_state,
+            OnComplete.DeleteApplication: self.delete_application,
+            OnComplete.NoOp: self.no_op,
+            OnComplete.OptIn: self.opt_in,
+            OnComplete.UpdateApplication: self.update_application,
+        }
+
+    def is_empty(self) -> bool:
+        for oc_action in self.dictify().values():
+            if oc_action.on_call is not None or oc_action.on_create is not None:
+                return False
+        return True
+
+
+OCActions.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class CondNode:
+    condition: Expr
+    branch: Expr
+
+
+CondNode.__module__ = "pyteal"
+
+
+@dataclass
+class CategorizedCondNodes:
+    method_calls_create: list[CondNode] = field(default_factory=list)
+    bare_calls_create: list[CondNode] = field(default_factory=list)
+    method_calls: list[CondNode] = field(default_factory=list)
+    bare_calls: list[CondNode] = field(default_factory=list)
+
+    def program_construction(self) -> Expr:
+        concatenated_ast = (
+            self.method_calls_create
+            + self.bare_calls_create
+            + self.method_calls
+            + self.bare_calls
+        )
+        if len(concatenated_ast) == 0:
+            raise TealInputError("ABIRouter: Cannot build program with an empty AST")
+        program: Cond = Cond(*[[n.condition, n.branch] for n in concatenated_ast])
+        return program
+
+
+CategorizedCondNodes.__module__ = "pyteal"
+
+
+class Router:
+    """
+    Class that help constructs:
+    - an ARC-4 app's approval/clear-state programs
+    - and a contract JSON object allowing for easily read and call methods in the contract
+    """
+
+    def __init__(
+        self,
+        name: str,
+        bare_calls: OCActions,
+    ) -> None:
+        """
+        Args:
+            name: the name of the smart contract, used in the JSON object.
+            bare_calls: the bare app call registered for each on_completion.
+        """
+
+        self.name: str = name
+        self.categorized_approval_ast = CategorizedCondNodes()
+        self.categorized_clear_state_ast = CategorizedCondNodes()
+        self.added_method_sig: set[str] = set()
+
+        self.__add_bare_call(bare_calls)
+
+    @staticmethod
+    def wrap_handler(
```

Comment:
@ahangsu Is `wrap_handler` intentionally a public method?

I'm wondering if it can be marked private by renmaing to `_wrap_handler`.  Intent is to limit the public API surface area so that we minimize chance of future breaking changes.

Once the API settles, I think it's generally worthwhile to consider visibility of all methods.  Once released as _public_, it's challenging to limit visibility.

Reply:
no, it was public for testing purpose, but we can make it hidden by leading underline. Good point 👍